en: Refactor backup restore (#1071)
pingcap/docs-tidb-operator: en/backup-restore-overview.md

---
title: Backup and Restore Overview

> summary:

Author: fixed in 4fd2af1

---

# Backup and Restore Overview

This document describes how to perform backup and restore on the TiDB cluster in Kubernetes. The backup and restore tools used are [BR](https://docs.pingcap.com/tidb/stable/backup-and-restore-tool), [Dumpling](https://docs.pingcap.com/tidb/stable/dumpling-overview), and [TiDB Lightning](https://docs.pingcap.com/tidb/stable/get-started-with-tidb-lightning).

TiDB Operator 1.1 and later versions implement the backup and restore methods using Custom Resource Definition (CRD):

+ If your TiDB cluster version is v3.1 or later, refer to the following documents:

>     - [Back up Data to S3-Compatible Storage Using BR](backup-to-aws-s3-using-br.md)

Reviewer: Two docs missed.

Author: fixed.

    - [Restore Data from S3-Compatible Storage Using BR](restore-from-aws-s3-using-br.md)
    - [Back up Data to GCS Using BR](backup-to-gcs-using-br.md)
    - [Restore Data from GCS Using BR](restore-from-gcs-using-br.md)

+ If your TiDB cluster version is earlier than v3.1, refer to the following documents:

    - [Back up Data to S3-Compatible Storage Using Dumpling](backup-to-s3.md)
    - [Restore Data from S3-Compatible Storage Using TiDB Lightning](restore-from-s3.md)
    - [Back up Data to GCS Using Dumpling](backup-to-gcs.md)

>     - [Restore Data from GCS Using TiDB Lightning](restore-from-gcs.md)

Author: Oh thanks! fixed in 4fd2af1

## Use scenarios

[Dumpling](https://docs.pingcap.com/tidb/stable/dumpling-overview) is a data export tool that exports data stored in TiDB/MySQL as SQL or CSV data files to get the logic backup. If you need to back up SST files (Key-Value pairs) directly or perform latency-insensitive incremental backup, refer to [BR](https://docs.pingcap.com/tidb/stable/backup-and-restore-tool). For real-time incremental backup, refer to [TiCDC](https://docs.pingcap.com/tidb/stable/ticdc-overview).

> [TiDB Lightning](https://docs.pingcap.com/tidb/stable/get-started-with-tidb-lightning) is a tool used for fast full data import into a TiDB cluster. TiDB Lightning supports Dumpling or CSV format data source. You can use TiDB Lightning for the following two purposes:

Author: fixed in 4fd2af1

    In extreme cases, if TiDB Operator fails to access the database, TiDB Operator cannot automatically recover the value of `tikv_gc_life_time` and treats the backup as failed. At this time, you can view `tikv_gc_life_time` of the current TiDB cluster using the following statement:

    {{< copyable "sql" >}}

    ```sql
    select VARIABLE_NAME, VARIABLE_VALUE from mysql.tidb where VARIABLE_NAME like "tikv_gc_life_time";
    ```

    In the output of the command above, if the value of `tikv_gc_life_time` is still larger than expected (10m by default), you need to [set `tikv_gc_life_time` back](https://docs.pingcap.com/tidb/stable/dumpling-overview#tidb-gc-settings-when-exporting-a-large-volume-of-data) to the previous value manually:

* `.spec.cleanPolicy`: The clean policy of the backup data when the backup CR is deleted.

    Three clean policies are supported:

    * `Retain`: On any circumstances, retain the backup data when deleting the backup CR.
    * `Delete`: On any circumstances, delete the backup data when deleting the backup CR.

>     * `OnFailure`: If the backup fails, delete the backup data when deleting the backup CR.

Author: - Thanks. I'll go with `under any circumstances`. (ref: https://www.collinsdictionary.com/zh/dictionary/english/under-any-circumstances-under-any-circumstance)
- It's already indented 4 spaces.

>     * `OnFailure`: If the backup fails, delete the backup data when deleting the backup CR.

Author: fixed in 4fd2af1

    ```bash
    tableFilter:
    - "*.*"
    - "!/^(mysql|test|INFORMATION_SCHEMA|PERFORMANCE_SCHEMA|METRICS_SCHEMA|INSPECTION_SCHEMA)$/.*"
    ```

    If the field is not configured, BR restores all the schemas in the backup file.

    > **Note:**
    >
    > To use the table filter to exclude `db.table`, you need to first add the `*.*` rule to include all tables. For example:
    >
    > ```
    > tableFilter:
    > - "*.*"
    > - "!db.table"
    > ```

* `.spec.br`: BR-related configuration. Refer to [BR fields](#br-fields).
* `.spec.s3`: S3-related configuration. Refer to [S3 storage fields](#s3-storage-fields).
* `.spec.gcs`: GCS-related configuration. Refer to [GCS fields](#gcs-fields).
* `.spec.local`: Persistent volume-related configuration. Refer to [Local storage fields](#local-storage-fields).

## BackupSchedule CR fields

The `backupSchedule` configuration consists of two parts. One is the unique configuration of `backupSchedule`, and the other is `backupTemplate`. `backupTemplate` specifies the configuration related to the cluster and remote storage, which is the same as the `spec` configuration of [the `Backup` CR](#backup-cr-fields).

> The following are the unique configuration items of `backupSchedule`:

Reviewer 2: ```suggestion
The unique configuration items of `backupSchedule` are as follows:
```

* `.spec.maxReservedTime`: A backup retention policy based on time. For example, if you set the value of this field to `24h`, only backup files within the recent 24 hours are retained. All backup files older than this value are deleted. For the time format, refer to [`func ParseDuration`](https://golang.org/pkg/time/#ParseDuration). If you have set `.spec.maxBackups` and `.spec.maxReservedTime` at the same time, the latter takes effect.
* `.spec.schedule`: The time scheduling format of Cron. Refer to [Cron](https://en.wikipedia.org/wiki/Cron) for details.
* `.spec.pause`: `false` by default. If this field is set to `true`, the scheduled scheduling is paused. In this situation, the backup operation will not be performed even if the scheduling time is reached. During this pause, the backup Garbage Collection runs normally. If you change `true` to `false`, the scheduled full backup process is restarted.

## Delete the Backup CR

You can delete the `Backup` CR or `BackupSchedule` CR by running the following commands:

{{< copyable "shell-regular" >}}

```shell
kubectl delete backup ${name} -n ${namespace}
kubectl delete backupschedule ${name} -n ${namespace}
```

> If you use TiDB Operator v1.1.2 or earlier versions, or if you use TiDB Operator v1.1.3 or later versions and set the value of `spec.cleanPolicy` to `Delete`, TiDB Operator deletes the backup data when it deletes the CR.

Reviewer 2: ```suggestion
If you use TiDB Operator v1.1.2 or an earlier version, or if you use TiDB Operator v1.1.3 or a later version and set the value of `spec.cleanPolicy` to `Delete`, TiDB Operator deletes the backup data when it deletes the CR.
```

In such cases, if you need to delete the namespace, it is recommended that you first delete all the `Backup`/`BackupSchedule` CRs and then delete the namespace.

> If you delete the namespace before you delete the `Backup`/`BackupSchedule` CR, TiDB Operator continues to create jobs to clean the backup data. However, since the namespace is in `Terminating` state, TiDB Operator fails to create a job, which causes the namespace to be stuck in the state.

Reviewer 2: ```suggestion
If you delete the namespace before you delete the `Backup`/`BackupSchedule` CR, TiDB Operator will keep creating jobs to clean the backup data. However, because the namespace is in `Terminating` state, TiDB Operator fails to create such a job, which causes the namespace to be stuck in this state.
```

TiDB Operator 1.1 and later versions implement the backup and restore methods using Custom Resource Definition (CRD):

+ If your TiDB cluster version is v3.1 or later, refer to the following documents:

    - [Back up Data to S3-Compatible Storage Using BR](backup-to-aws-s3-using-br.md)
    - [Back up Data to GCS Using BR](backup-to-gcs-using-br.md)
    - [Back up Data to PV Using BR](backup-to-pv-using-br.md)
    - [Restore Data from S3-Compatible Storage Using BR](restore-from-aws-s3-using-br.md)
    - [Restore Data from GCS Using BR](restore-from-gcs-using-br.md)
    - [Restore Data from PV Using BR](restore-from-pv-using-br.md)

+ If your TiDB cluster version is earlier than v3.1, refer to the following documents:

    - [Back up Data to S3-Compatible Storage Using Dumpling](backup-to-s3.md)
    - [Back up Data to GCS Using Dumpling](backup-to-gcs.md)
    - [Restore Data from S3-Compatible Storage Using TiDB Lightning](restore-from-s3.md)
    - [Restore Data from GCS Using TiDB Lightning](restore-from-gcs.md)

> ## Use scenarios

Reviewer 2: ```suggestion
## User scenarios
```
